aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1137,41 +1141,38 @@ Tensor _lu_solve_helper_cpu(const Tensor& self, const Tensor& LU_data, const Ten
-              "b should have 2 or 3 dimensions, but has ", self.dim(), " dimensions instead");
-  TORCH_CHECK(LU_data.dim() == 3,
-              "LU_data should have 3 dimensions, but has ", LU_data.dim(), " dimensions instead");
-  TORCH_CHECK(self.size(0) == LU_data.size(0),
-              "b and LU_data should have the same number of batches");
-  TORCH_CHECK(LU_pivots.size(1) == LU_data.size(2),
+  TORCH_CHECK(self.dim() >= 2,
+              "b should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
+  TORCH_CHECK(LU_data.dim() >= 2,
+              "LU_data should have at least 2 dimensions, but has ", LU_data.dim(), " dimensions instead");
+  TORCH_CHECK(LU_pivots.size(-1) == LU_data.size(-1),
               "Number of pivots per batch should be same as the dimension of the matrix");
-  TORCH_CHECK(LU_pivots.size(0) == LU_data.size(0),
-              "Batch dimensions of LU_pivots doesn't match batch dimensions of LU_data");
   TORCH_CHECK(LU_pivots.dtype() == at::kInt,
               "LU_pivots should be a Tensor of scalar type Int");
   TORCH_CHECK(LU_pivots.device() == LU_data.device(),
               "Expected LU_pivots and LU_data to be on the same device, "
               "but found LU_pivots on ", LU_pivots.device(), " and LU_data on ",
               LU_data.device(), " instead");
 
-  Tensor self_3D;
-  if (self.dim() == 2) {
-    TORCH_WARN("Passing RHS tensor with number of dimensions = 2 is deprecated, "
-               "and will be removed in the next release. Please unsqueeze the last dimension "
-               "to obtain an RHS tensor with number of right hand sides = 1");
-    self_3D = self.unsqueeze(2);
-  } else {
-    self_3D = self;
-  }
-  linearSolveCheckInputs(self_3D, LU_data, "lu_solve");
-
-  Tensor solution = at::_lu_solve_helper(self_3D, LU_data, LU_pivots);
-  if (self.dim() == 2) {
-    return solution.squeeze(2);
-  } else {
-    return solution;
-  }
+  // We check whether the batch dimensions of LU_pivots match the batch dimensions of LU_data
+  // e.g.: LU_pivots.sizes() = 4 x 3 x 2, LU_data.sizes() = 4 x 3 x 2 x 2 is a pair of correct inputs
+  // e.g.: LU_pivots.sizes() = 4 x 3 x 2, LU_data.sizes() = 12 x 2 x 2 is a pair of incorrect inputs
+  IntArrayRef pivots_sizes(LU_pivots.sizes().data(), LU_pivots.ndimension() - 1);
+  IntArrayRef lu_sizes(LU_data.sizes().data(), LU_data.ndimension() - 2);
+  TORCH_CHECK(pivots_sizes == lu_sizes,
+              "batch dimensions of LU_pivots doesn't match batch dimensions of LU_data");
+
+  Tensor self_broadcasted, LU_data_broadcasted;
+  std::tie(self_broadcasted, LU_data_broadcasted) = _linear_solve_broadcast_args(self, LU_data, "lu_solve");
+
+  // Now, we need to broadcast pivots too for the batch dimensions to match
+  IntArrayRef new_lu_sizes(LU_data_broadcasted.sizes().data(), LU_data_broadcasted.ndimension() - 2);
+  std::vector<int64_t> new_pivots_sizes = infer_size(new_lu_sizes, pivots_sizes);
```

Comment:
Hmm... is `new_pivots_sizes` just equal to (Python syntax) `LU_data_broadcasted.sizes()[:-2] + [LU_data.size(-1)]`? If so we can avoid the call into `infer_size`.

Reply:
Oh yes, why didn't I think of that?

In fact, there can be one other simplification: `new_pivots_sizes` is actually equal to `LU_data_broadcasted.sizes()[:-1]`

Reply:
Oh yeah, that is better